File: src/edge-agent/app/src/main/kotlin/com/fccmiddleware/edge/preauth/PreAuthStateMachine.kt
package com.fccmiddleware.edge.preauth

import com.fccmiddleware.edge.adapter.common.PreAuthStatus

object PreAuthStateMachine {
    val activeStatuses: Set<PreAuthStatus> = setOf(
        PreAuthStatus.PENDING,
        PreAuthStatus.AUTHORIZED,
        PreAuthStatus.DISPENSING,
    )

    val terminalStatuses: Set<PreAuthStatus> = setOf(
        PreAuthStatus.COMPLETED,
        PreAuthStatus.CANCELLED,
        PreAuthStatus.EXPIRED,
        PreAuthStatus.FAILED,
    )

    private val allowedTransitions: Map<PreAuthStatus, Set<PreAuthStatus>> = mapOf(
        PreAuthStatus.PENDING to setOf(
            PreAuthStatus.AUTHORIZED,
            PreAuthStatus.CANCELLED,
            PreAuthStatus.EXPIRED,
            PreAuthStatus.FAILED,
        ),
        PreAuthStatus.AUTHORIZED to setOf(
            PreAuthStatus.DISPENSING,
            PreAuthStatus.COMPLETED,
            PreAuthStatus.CANCELLED,
            PreAuthStatus.EXPIRED,
            PreAuthStatus.FAILED,
        ),
        PreAuthStatus.DISPENSING to setOf(
            PreAuthStatus.COMPLETED,
            PreAuthStatus.CANCELLED,
            PreAuthStatus.EXPIRED,
            PreAuthStatus.FAILED,
        ),
        PreAuthStatus.COMPLETED to emptySet(),
        PreAuthStatus.CANCELLED to emptySet(),
        PreAuthStatus.EXPIRED to emptySet(),
        PreAuthStatus.FAILED to emptySet(),
    )

    fun isActive(status: PreAuthStatus): Boolean = status in activeStatuses

    fun isTerminal(status: PreAuthStatus): Boolean = status in terminalStatuses

    fun canTransition(from: PreAuthStatus, to: PreAuthStatus): Boolean =
        allowedTransitions[from]?.contains(to) == true

    fun allowedTransitionsFrom(from: PreAuthStatus): Set<PreAuthStatus> =
        allowedTransitions[from].orEmpty()
}

File: src/desktop-edge-agent/src/FccDesktopAgent.Core/PreAuth/PreAuthStateMachine.cs
using FccDesktopAgent.Core.Adapter.Common;

namespace FccDesktopAgent.Core.PreAuth;

internal static class PreAuthStateMachine
{
    internal static IReadOnlyList<PreAuthStatus> ActiveStatuses { get; } =
    [
        PreAuthStatus.Pending,
        PreAuthStatus.Authorized,
        PreAuthStatus.Dispensing
    ];

    internal static IReadOnlyList<PreAuthStatus> TerminalStatuses { get; } =
    [
        PreAuthStatus.Completed,
        PreAuthStatus.Cancelled,
        PreAuthStatus.Expired,
        PreAuthStatus.Failed
    ];

    private static readonly IReadOnlyDictionary<PreAuthStatus, HashSet<PreAuthStatus>> AllowedTransitions =
        new Dictionary<PreAuthStatus, HashSet<PreAuthStatus>>
        {
            [PreAuthStatus.Pending] =
            [
                PreAuthStatus.Authorized,
                PreAuthStatus.Cancelled,
                PreAuthStatus.Expired,
                PreAuthStatus.Failed
            ],
            [PreAuthStatus.Authorized] =
            [
                PreAuthStatus.Dispensing,
                PreAuthStatus.Completed,
                PreAuthStatus.Cancelled,
                PreAuthStatus.Expired,
                PreAuthStatus.Failed
            ],
            [PreAuthStatus.Dispensing] =
            [
                PreAuthStatus.Completed,
                PreAuthStatus.Cancelled,
                PreAuthStatus.Expired,
                PreAuthStatus.Failed
            ],
            [PreAuthStatus.Completed] = [],
            [PreAuthStatus.Cancelled] = [],
            [PreAuthStatus.Expired] = [],
            [PreAuthStatus.Failed] = []
        };

    internal static IReadOnlyList<string> ActiveStatusNames { get; } =
        ActiveStatuses.Select(status => status.ToString()).ToArray();

    internal static bool IsActive(PreAuthStatus status) => ActiveStatuses.Contains(status);

    internal static bool IsTerminal(PreAuthStatus status) => TerminalStatuses.Contains(status);

    internal static bool CanTransition(PreAuthStatus from, PreAuthStatus to) =>
        AllowedTransitions.TryGetValue(from, out var allowed) && allowed.Contains(to);

    internal static IReadOnlyCollection<PreAuthStatus> GetAllowedTransitions(PreAuthStatus from) =>
        AllowedTransitions.TryGetValue(from, out var allowed)
            ? allowed
            : Array.Empty<PreAuthStatus>();
}
